test/adapter/json_api/errors_test.rb

```diff
@@ -0,0 +1,154 @@
+require 'test_helper'
+
+module ActiveModel
+  class Serializer
+    class Adapter
+      class JsonApi
+        class ErrorsTest < Minitest::Test
+
+          # see
+          # https://github.com/rails/rails/blob/4-2-stable/activemodel/lib/active_model/errors.rb
+          # The below allows you to do:
+          #
+          #   model = ModelWithErrors.new
+          #   model.validate!            # => ["cannot be nil"]
+          #   model.errors.full_messages # => ["name cannot be nil"]
+          class ModelWithErrors
+            # Required dependency for ActiveModel::Errors
+            extend ActiveModel::Naming
+            def initialize
+              @errors = ActiveModel::Errors.new(self)
+            end
+            attr_accessor :name
+            attr_reader   :errors
+
+            # The following methods are needed to be minimally implemented
+
+            def read_attribute_for_validation(attr)
+              send(attr)
+            end
+
+            def self.human_attribute_name(attr, options = {})
+              attr
+            end
+
+            def self.lookup_ancestors
+              [self]
+            end
+          end
+
+          def setup
+            @model = ModelWithErrors.new
+          end
+
+          class JsonApiErrorSerializer < ActiveModel::Serializer
+            attribute :source
+            attribute :detail
+
+            def source
+              {pointer: "data/attributes/#{object.first}"}
+            end
+
+            def detail
+              detail = object.last
+              if detail.respond_to?(:join)
+                detail.join
+              else
+                detail
+              end
+            end
+          end
+
+          class JsonApiErrorsSerializer < ActiveModel::Serializer
+            has_many :errors, serializer: JsonApiErrorSerializer
+
+            def errors
+              object.messages.to_a
+            end
+          end
+
+          class ActiveModel::Serializer::Adapter::JsonApiError < ActiveModel::Serializer::Adapter::FlattenJson
+          end
+
+          def test_active_model_errors
+            @model.errors.add(:name, "cannot be nil")
+            serializer_opts = {serializer: JsonApiErrorSerializer}
+            errors = @model.errors
+
+            serializer = JsonApiErrorSerializer.new(errors.first, serializer_opts)
+            expected_error_object =
+            {
+              source: { pointer: 'data/attributes/name' },
+              detail: 'cannot be nil'
+            }
```

Comment:
@dgeb Sorry to pull you in here, but I was wondering if you could help me clarify the expected [json-api response for multiple errors on the same field](http://jsonapi.org/format/#errors).

[Ember docs]( http://emberjs.com/blog/2015/06/18/ember-data-1-13-released.html#toc_using-json-api-error-object-format) make me think that multiple errors on the same field should be joined.  e.g.

old: 

```json
errors: {
  first_name: ['is too short', 'must contain an emoji']
}
```

new (one error object?):

```json
errors: [
  {
    source: { pointer: 'data/attributes/first_name' },
    detail: 'is too short AND must contain an emoji'
  }
]
```
or new (two error objects, as only the 'id' field must be unique across objects):

```json
errors: [
  {
    source: { pointer: 'data/attributes/first_name' },
    detail: 'is too short'
  },
  {
    source: { pointer: 'data/attributes/first_name' },
    detail: 'must contain an emoji'
  }
]
```

ref:
- https://github.com/json-api/json-api/pull/237 [specifically](https://github.com/json-api/json-api/commit/6901143531c14f41e95818455fe36228cebf6625) and https://github.com/json-api/json-api/pull/638
- https://github.com/json-api/json-api/pull/234
- https://github.com/json-api/json-api/issues/7
- https://github.com/json-api/json-api/pull/208